cip-ledger-state.md

```diff
@@ -0,0 +1,111 @@
+on the block chains and partly in what is referred to as ledger state (described more fully below).
+Extracting block chain data is relatively easy, but ledger state data is not. Currently these
+applications have to recreate and maintain ledger state themselves based on the block information
+they stream from the node over the local chain sync protocol. Recreation and maintenance of the
+ledger state is not only complex and a source of bugs but more importantly requires significant
+resources, specifically RAM. Ledger state in memory currently consumes 10 Gig of RAM and that is
+growing. In situations where the node and any such application run on the same machine, the machine
+ends up with twice the resource usage. The following proposal hopes to reduce resource usage and
+complexity for chain following applications like db-sync.
+
+# Current Situation
+
+Currently there is a local chain sync protocol which is really just the peer-to-peer protocol
+using a local domain socket rather than the TCP/IP socket normally used for P2P transport.
+
+The data transported over this local chain sync protocol is limited to block chain data. However, a
+Cardano node also maintains ledger state which includes:
+
+* The current UTxO state.
+* Current amount of ADA delegated to each stake pool.
+* Which stake address is currently delegated to each pool.
+* Rewards account balances for each stake address.
+* Current protocol parameters.
+
+The first of these ledger state components is by far the largest component and is probably not
+needed outside the node (and definitely not needed by db-sync). However the others are needed and
+stored by `cardano-db-sync` which gets these data sets by maintaining its own copy of ledger state
+and periodically extracting the parts required.
+
+This means that when `node` and `db-sync` are run on the same machine (which is the recommended
+configuration) that machine has two basically identical copies of ledger state. Ledger state is a
+*HUGE* data structure and the mainnet version currently consumes over 10 Gigabytes of memory.
+Furthermore, maintaining ledger state duplicates code functionality that is in `ouroboros-consensus`
+and the maintenance of ledger state has been the cause of about 90% of the bugs in `db-sync` over
+the last two years. The maintenance of ledger state also makes updating, running and maintaining
+`db-sync` by operators more difficult than it should be. Finally, if `db-sync` did not have to
+maintain ledger state, the size of the `db-sync` code base would probably decrease by about 50% and
+the bits removed are some of the most complicated parts.
+
+
+## Specification
+
+The proposed solution is an enhanced local chain sync protocol that would only be served over a
+local domain socket. The enhanced chain chain sync protocol would include the existing block chain
+data as well as events containing the ledger data that db-sync needs. This enhanced local chain
```

Comment:
This has always been said to be "impossible" because the ledger does not record past events and wouldn't be able to replay any of those events that are too far in the past. 

Wouldn't it make sense to start by _recording_ the events somewhere?

Reply:
That is effectively part of this proposal.